component/loki/source/kubernetes/kubetail/tailer.go

```diff
@@ -0,0 +1,323 @@
+package kubetail
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component/common/loki"
+	"github.com/grafana/agent/pkg/runner"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubetypes "k8s.io/apimachinery/pkg/types"
+)
+
+// tailerTask is the payload used to create tailers. It implements runner.Task.
+type tailerTask struct {
+	Options *Options
+	Target  *Target
+}
+
+var _ runner.Task = (*tailerTask)(nil)
+
+func (tt *tailerTask) Hash() uint64 { return tt.Target.Hash() }
+
+func (tt *tailerTask) Equals(other runner.Task) bool {
+	otherTask := other.(*tailerTask)
+
+	// Quick path: pointers are exactly the same.
+	if tt == otherTask {
+		return true
+	}
+
+	// Slow path: check individual fields which are part of the task.
+	return tt.Options == otherTask.Options &&
+		tt.Target.UID() == otherTask.Target.UID() &&
+		labels.Equal(tt.Target.Labels(), otherTask.Target.Labels())
+}
+
+// A tailer tails the logs of a Kubernetes container. It is created by a
+// [Manager].
+type tailer struct {
+	log    log.Logger
+	opts   *Options
+	target *Target
+
+	lset model.LabelSet
+}
+
+var _ runner.Worker = (*tailer)(nil)
+
+// newTailer returns a new Tailer which tails logs from the target specified by
+// the task.
+func newTailer(l log.Logger, task *tailerTask) *tailer {
+	return &tailer{
+		log:    log.WithPrefix(l, "target", task.Target.String()),
+		opts:   task.Options,
+		target: task.Target,
+
+		lset: newLabelSet(task.Target.Labels()),
+	}
+}
+
+func newLabelSet(l labels.Labels) model.LabelSet {
+	res := make(model.LabelSet, len(l))
+	for _, pair := range l {
+		res[model.LabelName(pair.Name)] = model.LabelValue(pair.Value)
+	}
+	return res
+}
+
+var retailBackoff = backoff.Config{
+	MinBackoff: time.Second,
+	MaxBackoff: time.Minute,
+}
+
+func (t *tailer) Run(ctx context.Context) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	level.Info(t.log).Log("msg", "tailer running")
+	defer level.Info(t.log).Log("msg", "tailer exited")
+
+	bo := backoff.New(ctx, retailBackoff)
+
+	handler := loki.NewEntryMutatorHandler(t.opts.Handler, func(e loki.Entry) loki.Entry {
+		// A log line got read, we can reset the backoff period now.
+		bo.Reset()
+		return e
+	})
+	defer handler.Stop()
+
+	for bo.Ongoing() {
+		err := t.tail(ctx, handler)
+		if err == nil {
+			terminated, err := t.containerTerminated(ctx)
+			if terminated {
+				// The container shut down and won't come back; we can stop tailing it.
+				return
+			} else if err != nil {
+				level.Warn(t.log).Log("msg", "could not determine if container terminated; will retry tailing", "err", err)
+			}
```

Comment:
I originally didn't have this logic, but had to implement it once I noticed that it kept retrying forever for init containers that stopped (and would never come back).

I briefly considered whether retrying forever was fine if the same log lines don't get sent more than once, but I'm concerned about how much extra pressure it would cause on the API server for clusters that make heavy use of init containers.